janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/LuceneIndex.java

```diff
@@ -330,23 +349,59 @@ private Document retrieveOrCreate(String docID, IndexSearcher searcher) throws I
         return doc;
     }
 
-    private void addToDocument(Document doc, List<IndexEntry> content, final KeyInformation.StoreRetriever information) {
+    private void addToDocument(Document doc, List<IndexEntry> content, final KeyInformation.StoreRetriever information, boolean isNew) {
         Preconditions.checkNotNull(doc);
 
         for (final IndexEntry e : content) {
             Preconditions.checkArgument(!e.hasMetaData(), "Lucene index does not support indexing meta data: %s", e);
             if (log.isTraceEnabled())
                 log.trace("Adding field [{}] on document [{}]", e.field, doc.get(DOCID));
 
-            if (doc.getField(e.field) != null) {
-                doc.removeFields(e.field);
+            KeyInformation ki = information.get(e.field);
+            if (!isNew) {
+                Iterator<IndexableField> it = doc.iterator();
+                while (it.hasNext()) {
```

Comment:
As for me, the logic is too complicated to be in one method. I would suggest to move logic under `if (!isNew)` into a separate method.

Reply:
Done